lib/application.js

```diff
@@ -398,85 +399,217 @@ app.enableAuth = function(options) {
     }
   };
 
-  this._verifyAuthModelRelations();
-
-  this.isAuthEnabled = true;
+  return this._verifyAuthModelRelations(app._warnOnBadAuthModelRelations)
```

Comment:
In existing applications that are not waiting for `enableAuth` to resolve, when the verification code throws an error, the rejected promise will end up as unhandled promise rejection. This typically produces only a console warning,  it does not crash the application as an uncaught error would.

```
(node:90355) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 3): Error
```

I am proposing to introduce an `options` flag to control the sync/async behaviour. For example:

```js
// old way - throws error
app.enableAuth({dataSource: 'db'});

// new way - returns rejected promise
app.enableAuth({dataSource: 'db', async: true})
  .then(...)
  .catch(...);

// alternatively, replace the flag with a new method name
app.enableAuthAsync({dataSource: 'db'});
```

(The "Async" prefix is already used by Bluebird's [promisify API](http://bluebirdjs.com/docs/api/promise.promisify.html))